lib/ansible/modules/packaging/os/package_facts.py

```diff
@@ -139,95 +150,275 @@
           }
         }
 '''
+import json
+
+from abc import ABCMeta, abstractmethod
+
+from ansible.module_utils._text import to_native, to_text
+from ansible.module_utils.basic import AnsibleModule, get_all_subclasses
+from ansible.module_utils.six import with_metaclass
+
+
+class TestFailed(Exception):
+    pass
+
+
+class PkgMgr(with_metaclass(ABCMeta, object)):
+
+    def __init__(self):
+
+        if not self.test():
+            raise TestFailed("Test failed to pass")
+
+    @abstractmethod
+    def test(self):
+        ''' This method is supposed to return True/False if the package manager is installed/usable by this module '''
+        pass
+
+    @abstractmethod
+    def list_installed(self):
+        ''' This method should return a list of installed packages, each list item will be passed to get_package_details '''
+        pass
+
+    @abstractmethod
+    def get_package_details(self, package):
+        ''' This takes a 'package' item and returns a dictionary with the package information, name and version are minimal requirements '''
+        pass
+
+    def get_packages(self):
+        ''' Take all of the above and return a dictionary of lists of dictionaries (package = list of installed versions) '''
+
+        installed_packages = {}
+        for package in self.list_installed():
+            package_details = self.get_package_details(package)
+            package_details['source'] = self.__class__.__name__.lower()
+            name = package_details['name']
+            if name not in installed_packages:
+                installed_packages[name] = [package_details]
+            else:
+                installed_packages[name].append(package_details)
+        return installed_packages
+
+
+class LibMgr(PkgMgr):
+
+    LIB = None
```

Comment:
Also, attributes deserve to have their own docstrings:
```suggestion
    lib_name = None
    """Name of importable used to access the current package manager."""
```

Reply:
Docstrings for attributes should be formatted as:
``` python
#: Import name of library for interacting with the package manager
@abstractproperty
def lib_name(self):
    pass
```

http://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoattribute

Reply:
No, that's not attribute docstrings.

PEP 257 says:
> String literals occurring immediately after a simple assignment at the top level of a module, class, or __init__ method are called "attribute docstrings".

That's my intention. They look like in code and they shouldn't be comments.


Your own link to sphinx docs says that it supports this style as well:
> For module data members and class attributes, documentation can [...], **or in a docstring** _after_ the definition.

And then it gives example:
```python
class Foo:
    """Docstring for class Foo."""

    [...]

    baz = 2
    """Docstring for class attribute Foo.baz."""

    def __init__(self):
        [...]

        self.spam = 4
        """Docstring for instance attribute spam."""
```

Reply:
Ah, I remember discussing htis with someone else many years ago....

Currently there are no attribute docstrings in python.  So my preference is that we use the comment form rather than the string literal form.  Comments pose a lower overhead to the interpreter (they are ignored while string literals are parsed, processed, and objects created) and since the string literal form is ultimately discarded because they aren't assigned to any variable names it makes sense to force python to do as little work as possible for these.